### DISCUSSION TITLE
Help correcting MD5 in ETag

### DISCUSSION THREAD

**Discussion by audouts** [minio/minio#17146] [Q&A] [answered]
I'm using `mc mv` to add files to a bucket and noticed that it's putting something weird in the ETag header. It looks like an MD5 with a hyphenated value at the end. What is that format?

I tried using `--disable-multipart` and that fixes the issue. I need the standard MD5 ETag.

How can I fix all the files that have already been moved?

**Comment by klauspost** [comment]
Your assumption is wrong. Etags are not MD5 - but may be.

Multipart e-tags are not MD5, but a an MD5 of MD5s of each part.

Same for encrypted objects.

**Reply by audouts** [reply]
Yes, I know they are not MD5 hashes by definition. I did not assume they are; I asked what the format was that I didn't recognize.

The standard value being put there by MinIO is an MD5 for single-part files. I would still like those files converted back to single part and have an MD5 in the ETag header.

**Reply by klauspost** [reply]
@audouts That is not how it works. The `-X` indicates that it is a multipart and there is no conversion. The Etag is the etag and not the MD5.

Your beef is with the AWS S3 designers. There is nothing we can do.

**Reply by audouts** [reply]
It's not a beef. I am asking for help. If you cannot or do not want to explain the meaning of the ETag or offer any suggestions for my situation, that's fine but your posts are not helpful.

**Reply by harshavardhana** [reply, marked as answer]
ETag has multiple meanings and its easy to confuse them to be md5sum - https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html

Here is the explanation that you might want @audouts 

```
ETag
The entity tag is a hash of the object. The ETag reflects changes only to the contents of an object, not its metadata. The ETag may or may not be an MD5 digest of the object data. Whether or not it is depends on how the object was created and how it is encrypted as described below:

- Objects created by the PUT Object, POST Object, or Copy operation, or through the AWS Management Console, and are encrypted by SSE-S3 or plaintext, have ETags that are an MD5 digest of their object data.

- Objects created by the PUT Object, POST Object, or Copy operation, or through the AWS Management Console, and are encrypted by SSE-C or SSE-KMS, have ETags that are not an MD5 digest of their object data.

- If an object is created by either the Multipart Upload or Part Copy operation, the ETag is not an MD5 digest, regardless of the method of encryption. If an object is larger than 16 MB, the AWS Management Console will upload or copy that object as a Multipart Upload, and therefore the ETag will not be an MD5 digest.
```

This kind of confusing implementation is an AWS S3 baggage sadly we have to carry :-)

**Reply by audouts** [reply]
Thank you @harshavardhana. I still wonder how the ETag value is generated in this scenario:

_If an object is created by either the Multipart Upload or Part Copy operation, the ETag is not an MD5 digest_

It's not important but I am curious.

As for my situation, I guess the only option is to delete all the objects and re-copy them with multi-part disabled.

**Reply by harshavardhana** [reply]
It is md5sum(md5-part1.. md5-partN)-N

**Reply by audouts** [reply]
Ah, thank you. :)

**Reply by klauspost** [reply]
Note that MinIO now supports [checksums](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html#large-object-checksums).
